File: src/Recursion/SubsetSum.java
package Recursion;

import java.util.ArrayList;
import java.util.Collections;

public class SubsetSum {
    static void func(int ind,
                     int sum,
                     ArrayList<Integer> arr,
                     int N,
                     ArrayList<Integer> sumSubset) {
        if (ind == N) {
            sumSubset.add(sum);
            return;
        }

        // pick the element
        func(ind + 1, sum + arr.get(ind), arr, N, sumSubset);

        // Do-not pick the element
        func(ind + 1, sum, arr, N, sumSubset);
    }

    public static ArrayList<Integer> subsetSums(ArrayList<Integer> arr,
                                                int N) {
        ArrayList<Integer> sumSubset = new ArrayList<>();
        func(0, 0, arr, N, sumSubset);
        Collections.sort(sumSubset);
        return sumSubset;
    }

    public static void main(String[] args) {

        ArrayList<Integer> arr = new ArrayList<>();
        arr.add(1);
        arr.add(2);
        arr.add(3);
        arr.add(4);
        arr.add(5);
        System.out.println(subsetSums(arr, 5));


    }

    static int subset(int[] arr,
                      int n,
                      int i,
                      int sum) {
        //base
        if (i == n) {
            if (sum == 0) return 1;
            return 0;
        }
        int inc = subset(arr, n, i + 1, sum - arr[i]);
        int ex = subset(arr, n, i + 1, sum);
        return ex + inc;
    }

}

File: src/recursionCpp/subsetSum.cpp
#include <bits/stdc++.h>
using namespace std;
 

void func(int ind, int sum, vector<int> &arr, int N, vector<int> &sumSubset)
{

    if (ind == N)
    {
        sumSubset.push_back(sum);
        cout<<sum<<endl;
        return;
    }

    func(ind + 1, sum + arr[ind], arr, N, sumSubset);
    func(ind + 1, sum, arr, N, sumSubset);
}
vector<int> subsetSum(vector<int> &arr, int N, int sum)
{
    vector<int> sumSubset;
    func(0, 0, arr, N, sumSubset);
    sort(sumSubset.begin(), sumSubset.end());
    return sumSubset;
}
int main()
{
    vector<int> arr = {1, 2, 3};
    int N = arr.size();
    int sum = 10;
    vector<int> sumSubset = subsetSum(arr, N, sum);
    for (auto x : sumSubset)
        cout << x << " ";
    cout << endl;
}